contrib/plugins/ceph/ceph.go

```diff
@@ -356,6 +391,136 @@ func (imgMgr *ImageMgr) GetSnapshots() (*[]SnapshotResponse, error) {
+	if cap == "0" {
+		return 0
+	}
+	UnitMapper := map[string]uint64{
+		"K": 0, //shift bit
+		"M": 10,
+		"G": 20,
+		"T": 30,
+		"P": 40,
+	}
+	unit := strings.ToUpper(cap[len(cap)-1:])
+	num, err := strconv.ParseInt(cap[:len(cap)-1], 10, 64)
+	if err != nil {
+		log.Println("[Error] Cannot convert this number", err)
+		return 0
+	}
+	if val, ok := UnitMapper[unit]; ok {
+		return num << val >> SIZE_SHIFT_BIT
+	} else {
+		log.Println("[Error] strage unit is not found.")
+		return 0
+	}
+}
+
+func (imgMgr *ImageMgr) getPoolsCapInfo() ([][]string, error) {
+	const poolStartLine = 5
+	output, err := execCmd("ceph df")
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+	lines := strings.Split(output, "\n")
+	var poolsInfo [][]string
+	for i := poolStartLine; i < len(lines); i++ {
+		poolsInfo = append(poolsInfo, strings.Fields(lines[i]))
+	}
+	return poolsInfo, nil
+}
+
+func (imgMgr *ImageMgr) getGlobalCapInfo() ([]string, error) {
+	const globalCapInfoLine = 2
+	output, err := execCmd("ceph df")
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+	lines := strings.Split(output, "\n")
+	return strings.Fields(lines[globalCapInfoLine]), nil
+}
+
+func (imgMgr *ImageMgr) getPoolsAttr() (map[string][]string, error) {
+	cmd := "ceph osd pool ls detail | grep \"^pool\"| awk '{print $3, $4, $6, $10}'"
+	output, err := execCmd(cmd)
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+	lines := strings.Split(output, "\n")
+	var poolDetail = make(map[string][]string)
+	for i := range lines {
+		if lines[i] == "" {
+			continue
+		}
+		str := strings.Fields(lines[i])
+		key := strings.Replace(str[0], "'", "", -1)
+		val := str[1:]
+		poolDetail[key] = val
+	}
+	return poolDetail, nil
+}
+
+func (imgMgr *ImageMgr) getPoolParam(line []string) map[string]interface{} {
+	param := make(map[string]interface{})
+	param["redundancyType"] = line[POOL_TYPE]
+	if param["redundancyType"] == "replicated" {
+		param["replicateSize"] = line[POOL_TYPE_SIZE]
+	} else {
+		param["erasureSize"] = line[POOL_TYPE_SIZE]
+	}
+	param["crushRuleset"] = line[POOL_CRUSH_RULESET]
+	return param
+}
+
+func (imgMgr *ImageMgr) ListPools() (*[]PoolResponse, error) {
+	pc, err := imgMgr.getPoolsCapInfo()
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+	gc, err := imgMgr.getGlobalCapInfo()
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+	pa, err := imgMgr.getPoolsAttr()
+	if err != nil {
+		log.Println("[Error]:", err)
+		return nil, err
+	}
+
+	var pools []PoolResponse
+	for i := range pc {
+		name := pc[i][POOL_NAME]
+		param := imgMgr.getPoolParam(pa[name])
+		totalCap := imgMgr.parseCapStr(gc[GLOBAL_SIZE])
+		maxAvailCap := imgMgr.parseCapStr(pc[i][POOL_MAX_AVAIL])
+		availCap := imgMgr.parseCapStr(gc[GLOBAL_AVAIL])
+		pool := PoolResponse{
+			Name:       name,
+			ID:         uuid.NewV5(uuid.NamespaceOID, name).String(),
```

Comment:
Looks great, but one thing need to be careful here is we used to uuid.NewV4() to create uuid, why using NewV5() here? Could it cause some hidden bugs?

Reply:
@leonwanghui The UUIDV5 is not upgrade version of UUID V4. Their main difference is the implementation of the algorithm. Version 5  UUID use the SHA1 hash functions , to combine a namespace with a piece of already unique data to generate a UUID. Version 4 UUIDs essentially use just 16 bytes of randomness  to generate a UUID.More detail infomation you can look at [here](http://stackoverflow.com/questions/703035/when-are-you-truly-forced-to-use-uuid-as-part-of-the-design/786541#786541). I use UUIDv5 just because I want to use ceph pool name to  generate the UUIDs so that UUID will be a fix value.

Reply:
ok, great :p